File: services/docreader/parser/output_inventory.py
"""Lossless export inventory profile of the shared DocReader service.

Text/grids use the existing parser DTOs. The versioned metadata sidecar records
physical carriers and explicit omissions; it never substitutes for source text.
"""
import base64
import hashlib
import json
import re
from importlib.metadata import version
from collections import Counter
from io import BytesIO
from pathlib import Path
from types import SimpleNamespace

from docx import Document as DocxDocument
from docx.oxml import parse_xml
from docx.oxml.ns import qn
from docx.table import Table
from docx.text.paragraph import Paragraph

from docreader.models.document import (
    AttachmentLocator, Document, DocumentLocator, ParagraphImageParent,
    StructuredSourceUnit, StructuredSourceUnitKind, TableCellImageParent,
)
from docreader.parser.docx_parser import (
    _docx_package_image_payloads, _docx_table_anchors, _drawing_units,
)

PROFILE = "output_inventory_v1"


def _entry(unit, part, ordinal, kind, *, bookmarks=None, fields=None, reason=None,
           heading_level=None, field_region=None):
    return {
        "unit_key": unit.key, "part": part, "ordinal": ordinal, "kind": kind,
        "bookmarks": bookmarks or [], "fields": fields or [],
        "heading_level": heading_level, "field_region": field_region,
        "status": "not_checked" if reason else "extracted", "reason": reason,
    }


def _style_names(doc):
    """styleId → `w:name`. Editors renumber `w:styleId` (Word writes plain digits
    after a round trip), so an outline level may only be read from `w:name`."""
    names = {}
    for style in doc.styles.element.iter(qn("w:style")):
        identity = style.get(qn("w:styleId"))
        label = style.find(qn("w:name"))
        if identity and label is not None:
            names[identity] = label.get(qn("w:val")) or ""
    return names


def _heading_level(child, style_names):
    properties = child.find(qn("w:pPr"))
    reference = properties.find(qn("w:pStyle")) if properties is not None else None
    identity = reference.get(qn("w:val")) if reference is not None else None
    match = re.match(r"^heading\s+(\d+)$", (style_names.get(identity or "") or "").strip(), re.IGNORECASE)
    if match is None:
        return None
    try:
        return max(1, int(match.group(1)))
    except ValueError:
        return None


def _field_region(child, stack):
    """Classify a story child against enclosing field regions.

    A table of contents spans several paragraphs: `fldChar begin` opens it, the
    entries follow as ordinary paragraphs, `fldChar end` closes it. Those entries
    repeat chapter titles verbatim, so a reader that cannot see the region reads
    every chapter twice. `stack` carries the open regions across children.
    """
    region = stack[-1] if stack else None
    for node in child.iter(qn("w:fldChar"), qn("w:instrText"), qn("w:fldSimple")):
        if node.tag == qn("w:fldChar"):
            kind = node.get(qn("w:fldCharType"), "")
            if kind == "begin":
                stack.append("field")
                region = region or "field"
            elif kind == "end" and stack:
                region = region or stack[-1]
                stack.pop()
        elif node.tag == qn("w:instrText"):
            if (node.text or "").strip().upper().startswith("TOC"):
                if stack:
                    stack[-1] = "toc"
                region = "toc"
        elif node.get(qn("w:instr"), "").strip().upper().startswith("TOC"):
            region = "toc"
        else:
            region = region or "field"
    return region


def _plain_picture_is_backed(drawing, owner, image_refs):
    """Only remove a redundant carrier when it is exactly a retained picture.

    Mixed graphics, linked images, crop/effect/rotation carriers remain separate
    omissions. A nearby successfully decoded picture is never enough.
    """
    data = drawing.xpath(".//*[local-name()='graphicData']")
    picture_uri = "http://schemas.openxmlformats.org/drawingml/2006/picture"
    if len(data) != 1 or data[0].get("uri") != picture_uri:
        return False
    if len(data[0]) != 1 or data[0][0].tag != f"{{{picture_uri}}}pic":
        return False
    blips = drawing.xpath(".//*[local-name()='blip']")
    if len(blips) != 1 or len(blips[0]):
        return False
    if drawing.xpath(".//*[local-name()='txbxContent' or local-name()='effectLst' or local-name()='effectDag']"):
        return False
    for rect in drawing.xpath(".//*[local-name()='srcRect']"):
        if any(value != "0" for value in rect.attrib.values()):
            return False
    for transform in drawing.xpath(".//*[local-name()='xfrm']"):
        if any(transform.get(name, "0") not in {"0", "false", "off"} for name in ("rot", "flipH", "flipV")):
            return False
    relation = blips[0].get(qn("r:embed"))
    part = owner.part.related_parts.get(relation) if relation else None
    raw = getattr(part, "blob", None)
    if not isinstance(raw, bytes) or not raw:
        return False
    extension = Path(str(getattr(part, "partname", ""))).suffix.lower() or ".bin"
    return f"images/{hashlib.sha256(raw).hexdigest()}{extension}" in image_refs


def _docx(content):
    doc = DocxDocument(BytesIO(content))
    style_names = _style_names(doc)
    units, entries = [], []
    image_ordinal = 0
    parts = {str(part.partname): part for part in doc.part.package.parts}
    stories = [(doc.part, doc.element.body)]
    used_parts = {str(doc.part.partname)}

    def unsupported(part, ordinal, reason):
        unit = StructuredSourceUnit(
            key=f"carrier:{len(units)}", ordinal=len(units),
            kind=StructuredSourceUnitKind.ATTACHMENT_REGION, text="",
            locator=AttachmentLocator(part_name=part, relationship_type="output_carrier"),
        )
        units.append(unit)
        entries.append(_entry(unit, part, ordinal, "not_checked", reason=reason))

    # Resolve inheritance without creating missing header/footer definitions.
    # A first/even story becomes applicable only when that section enables it.
    references = {}
    used_references = []
    even_pages = doc.settings.odd_and_even_pages_header_footer
    for section in doc.element.xpath(".//w:sectPr"):
        for reference in section:
            if reference.tag in {qn("w:headerReference"), qn("w:footerReference")}:
                references[(reference.tag, reference.get(qn("w:type"), "default"))] = reference.get(qn("r:id"))
        enabled = {"default"}
        title_page = section.find(qn("w:titlePg"))
        if title_page is not None and title_page.get(qn("w:val"), "true") not in {"0", "false", "off"}:
            enabled.add("first")
        if even_pages:
            enabled.add("even")
        used_references.extend(relation for (_, story_type), relation in references.items() if story_type in enabled)
    for relation in dict.fromkeys(used_references):
        part = doc.part.related_parts.get(relation)
        if part is None:
            unsupported(str(doc.part.partname), len(entries), f"missing story relationship: {relation}")
            continue
        name = str(part.partname)
        if name not in used_parts:
            used_parts.add(name)
            stories.append((part, parse_xml(part.blob)))

    for kind in ("footnote", "endnote"):
        referenced = {node.get(qn("w:id")) for _, root in stories for node in root.iter(qn(f"w:{kind}Reference"))}
        if not referenced:
            continue
        candidates = [part for rel in doc.part.rels.values()
                      if rel.reltype.endswith(f"/{kind}s") and not rel.is_external
                      for part in [rel.target_part]]
        found = set()
        for part in candidates:
            used_parts.add(str(part.partname))
            root = parse_xml(part.blob)
            for note in root:
                identity = note.get(qn("w:id"))
                if identity in referenced:
                    found.add(identity)
                    stories.append((part, note))
        for missing in sorted(referenced - found):
            unsupported(str(doc.part.partname), len(entries), f"missing {kind}: {missing}")

    part_ordinals = {}
    for part, story in stories:
        name = str(part.partname)
        owner = SimpleNamespace(part=part)
        # Bookmarks that wrap a whole paragraph sit as siblings of `w:p`. Word
        # normalizes them inline, but a freshly compiled document still has them
        # outside, so they belong to the next carrier rather than to nothing.
        block_bookmarks = []
        open_fields = []
        for child in story:
            if child.tag in {qn("w:bookmarkStart"), qn("w:bookmarkEnd")}:
                bookmark = child.get(qn("w:name"))
                if bookmark:
                    block_bookmarks.append(bookmark)
                continue
            ordinal = part_ordinals.get(name, 0)
            part_ordinals[name] = ordinal + 1
            bookmarks = block_bookmarks + [
                node.get(qn("w:name")) for node in child.iter(qn("w:bookmarkStart"))
                if node.get(qn("w:name"))
            ]
            block_bookmarks = []
            fields = [node.text or "" for node in child.iter(qn("w:instrText"))]
            fields += [node.get(qn("w:instr"), "") for node in child.iter(qn("w:fldSimple"))]
            field_region = _field_region(child, open_fields)
            section = len(units)
            locator = DocumentLocator(section_ordinal=section, heading_path="")
            grid = None
            anchors = []
            heading_level = None
            if child.tag == qn("w:p"):
                text = Paragraph(child, owner).text
                heading_level = _heading_level(child, style_names)
                kind = "paragraphs"
            elif child.tag == qn("w:tbl"):
                try:
                    grid, anchors = _docx_table_anchors(Table(child, owner))
                except (ValueError, TypeError) as error:
                    unsupported(name, ordinal, f"unsupported table geometry: {error}")
                    continue
                locator.table_ordinal = ordinal
                text, kind = "", "table"
            elif child.tag == qn("w:sectPr"):
                text, kind = "", "section"
            else:
                unsupported(name, ordinal, f"unsupported story carrier: {child.tag}")
                continue
            unit = StructuredSourceUnit(
                key=f"story:{len(units)}", ordinal=len(units),
                kind=StructuredSourceUnitKind.TABLE_REGION if grid else StructuredSourceUnitKind.SECTION,
                text=text, locator=locator, grid=grid,
            )
            units.append(unit)
            entries.append(_entry(unit, name, ordinal, kind, bookmarks=bookmarks, fields=fields,
                                  heading_level=heading_level, field_region=field_region))
            # Preserve image bytes and their established typed parent identity.
            before = len(units)
            if grid:
                for row, column, cell in anchors:
                    image_ordinal = _drawing_units(cell, owner, units, image_ordinal,
                        TableCellImageParent(section_ordinal=section, table_ordinal=ordinal,
                                             row_ordinal=row, cell_ordinal=column))
            else:
                image_ordinal = _drawing_units(child, owner, units, image_ordinal,
                    ParagraphImageParent(section_ordinal=section, paragraph_ordinal=ordinal))
            extracted_image_refs = {image.locator.original_ref for image in units[before:]}
            for image in units[before:]:
                entries.append(_entry(image, name, ordinal, "image", reason="image requires visual review"))
            # Never silently treat unsupported XML carriers as ordinary text.
            unsupported_tags = {qn(f"w:{tag}") for tag in (
                "drawing", "pict", "object", "txbxContent", "altChunk", "sdt", "del", "ins",
            )}
            for descendant in child.iter():
                if descendant.tag in unsupported_tags:
                    if descendant.tag == qn("w:drawing") and _plain_picture_is_backed(descendant, owner, extracted_image_refs):
                        continue
                    unsupported(name, ordinal, f"visual or unsupported nested carrier: {descendant.tag}")
                elif descendant.tag == qn("w:tbl") and descendant is not child:
                    unsupported(name, ordinal, "nested table requires independent geometry review")
            if fields:
                unsupported(name, ordinal, "field instruction/result consistency requires review")
        if block_bookmarks:
            unsupported(name, part_ordinals.get(name, 0),
                        f"block bookmarks carry no following content: {block_bookmarks}")

    bookmark_counts = Counter(name for entry in entries for name in entry["bookmarks"])
    for entry in list(entries):
        duplicate = [name for name in entry["bookmarks"] if bookmark_counts[name] > 1]
        if duplicate:
            unsupported(entry["part"], entry["ordinal"], f"duplicate bookmarks: {duplicate}")

    # Other content-bearing relationship parts are explicit omissions, not a
    # generic scan of every XML part (styles and settings are not body text).
    for part in parts.values():
        for relation in part.rels.values():
            leaf = relation.reltype.rsplit("/", 1)[-1]
            if leaf not in {"aFChunk", "oleObject", "package", "chart", "diagramData", "comments"}:
                continue
            target = relation.target_ref if relation.is_external else str(relation.target_part.partname)
            unsupported(str(part.partname), len(entries), f"unsupported {leaf} relationship: {target}")

    return Document(content="\n".join(unit.text for unit in units),
                    images=_docx_package_image_payloads(content), structured_source_units=units), entries


def parse_output_inventory(file_name, file_type, content):
    media = file_type.lower().lstrip(".")
    if media == "docx":
        document, entries = _docx(content)
        parser = f"docreader-output-inventory-v1/python-docx/{version('python-docx')}"
        config = {"profile": PROFILE, "preserve_story_occurrences": True}
    elif media == "pdf":
        from docreader.parser.pdf_parser import PDFParser
        document = PDFParser(file_name=file_name, file_type="pdf", output_inventory=True).parse_into_text(content)
        entries = []
        for unit in document.structured_source_units:
            page = unit.locator.page_ordinal
            image = unit.kind == StructuredSourceUnitKind.IMAGE_REGION
            entries.append(_entry(unit, f"pdf:page:{page + 1}", unit.ordinal,
                                  "image" if image else "table" if unit.grid else "pdf_page",
                                  reason="image requires visual review" if image else None))
        parser = f"docreader-output-inventory-v1/pdfium/{version('pypdfium2')}"
        config = {"profile": PROFILE, "preserve_all_pages": True, "preserve_page_images": True, "text_cleaning": False}
    else:
        raise ValueError("output inventory supports exact DOCX and PDF bytes only")
    # Freeze the actual implementation and dependency identities, not just a
    # manually incremented profile label. Config records all extraction knobs.
    from docreader.config import CONFIG
    from docreader.parser import pdf_parser
    implementation = hashlib.sha256()
    for filename in ("output_inventory.py", "docx_parser.py", "pdf_parser.py", "pdf_tables.py"):
        implementation.update(Path(__file__).with_name(filename).read_bytes())
    parser += f"/implementation/{implementation.hexdigest()}/lxml/{version('lxml')}/pillow/{version('Pillow')}"
    config["settings"] = {} if media == "docx" else {
        "pdf_render_dpi": CONFIG.pdf_render_dpi, "pdf_jpeg_quality": CONFIG.pdf_jpeg_quality,
        "pdf_render_max_edge": CONFIG.pdf_render_max_edge,
        "layout_ordering": pdf_parser.LAYOUT_ORDERING,
        "extract_embedded_images": pdf_parser.EXTRACT_EMBEDDED_IMAGES,
    }
    image_sha256 = {ref: hashlib.sha256(base64.b64decode(data)).hexdigest()
                    for ref, data in (document.images or {}).items()}
    manifest = {
        "schema_version": 1, "profile": PROFILE,
        "file_sha256": hashlib.sha256(content).hexdigest(), "parser": parser,
        "config": config, "image_sha256": image_sha256, "page_count": document.metadata.get("page_count"), "units": entries,
    }
    document.metadata["output_inventory_manifest"] = json.dumps(manifest, ensure_ascii=False, separators=(",", ":"))
    return document

File: crates/docparser/src/output_inventory.rs
//! Explicit, fail-closed final-file profile over the shared DocReader transport.
use crate::{
    ConvertError, DocReaderReadError, ReadResult, StructuredSourceLocator, StructuredSourceUnitKind,
};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::collections::{BTreeMap, BTreeSet, HashMap};
use tokio_util::sync::CancellationToken;

pub const OUTPUT_INVENTORY_PROFILE: &str = "output_inventory_v1";

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct OutputTableLayout {
    pub widths_twips: Vec<u32>,
    pub header_rows: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct OutputInventoryEntry {
    pub unit_key: String,
    pub part: String,
    pub ordinal: usize,
    pub kind: String,
    pub bookmarks: Vec<String>,
    pub fields: Vec<String>,
    /// Outline level of a heading paragraph, read from the style `w:name` only:
    /// editors renumber `w:styleId` after a round trip.
    #[serde(default)]
    pub heading_level: Option<u32>,
    /// Paragraph style `w:name` when present. Fill only accepts compiler-safe
    /// styles; custom names become not_checked.
    #[serde(default)]
    pub style_name: Option<String>,
    /// Set when the carrier sits inside a field region. Table-of-contents
    /// entries repeat chapter titles verbatim, so a reader that cannot see the
    /// region counts every chapter twice.
    #[serde(default)]
    pub field_region: Option<String>,
    pub table_layout: Option<OutputTableLayout>,
    pub status: String,
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct OutputInventoryManifest {
    pub schema_version: u32,
    pub profile: String,
    pub file_sha256: String,
    pub parser: String,
    pub config: Value,
    pub page_count: Option<u32>,
    pub image_sha256: BTreeMap<String, String>,
    pub units: Vec<OutputInventoryEntry>,
}

pub struct OutputInventoryRead {
    pub parsed: ReadResult,
    pub manifest: OutputInventoryManifest,
}

pub async fn read_output_inventory(
    file_name: &str,
    bytes: Vec<u8>,
    cancel: &CancellationToken,
) -> Result<OutputInventoryRead, DocReaderReadError> {
    let file_type = file_name.rsplit('.').next().unwrap_or("");
    if !matches!(file_type, "docx" | "pdf") {
        return Err(DocReaderReadError::Configuration(
            "output inventory requires DOCX or PDF".into(),
        ));
    }
    let expected = hex::encode(Sha256::digest(&bytes));
    let overrides = HashMap::from([("output_inventory".into(), "v1".into())]);
    let parsed = crate::grpc::read_classified(
        crate::grpc::ConvertRequest {
            file_content: bytes,
            file_name: file_name.into(),
            file_type: file_type.into(),
            url: String::new(),
            title: file_name.into(),
            parser_engine: "builtin".into(),
            parser_engine_overrides: overrides,
        },
        cancel,
    )
    .await?;
    let manifest = validate_output_inventory(&parsed, &expected, file_type)
        .map_err(|error| DocReaderReadError::InvalidResponse(error.0))?;
    Ok(OutputInventoryRead { parsed, manifest })
}

/// Also usable with a recorded service response; no local parsing fallback.
pub fn validate_output_inventory(
    parsed: &ReadResult,
    expected_sha256: &str,
    file_type: &str,
) -> Result<OutputInventoryManifest, ConvertError> {
    let invalid = |message: &str| ConvertError(format!("DocReader output inventory: {message}"));
    if !parsed.error.is_empty() {
        return Err(invalid(&parsed.error));
    }
    let manifest: OutputInventoryManifest = serde_json::from_str(
        parsed
            .metadata
            .get("output_inventory_manifest")
            .ok_or_else(|| invalid("profile receipt missing; service upgrade required"))?,
    )
    .map_err(|e| invalid(&e.to_string()))?;
    if manifest.schema_version != 1
        || manifest.profile != OUTPUT_INVENTORY_PROFILE
        || manifest.config["profile"] != OUTPUT_INVENTORY_PROFILE
        || manifest.file_sha256 != expected_sha256
        || manifest.parser.trim().is_empty()
    {
        return Err(invalid("profile or immutable file identity mismatch"));
    }
    let expected_config = match file_type {
        "docx" => {
            serde_json::json!({"profile":OUTPUT_INVENTORY_PROFILE,"preserve_story_occurrences":true})
        }
        "pdf" => {
            serde_json::json!({"profile":OUTPUT_INVENTORY_PROFILE,"preserve_all_pages":true,"preserve_page_images":true,"text_cleaning":false})
        }
        _ => return Err(invalid("unsupported file type")),
    };
    let mut checked_config = manifest.config.clone();
    let settings = checked_config
        .as_object_mut()
        .and_then(|config| config.remove("settings"));
    if !settings.is_some_and(|settings| settings.is_object()) {
        return Err(invalid("parser settings receipt missing"));
    }
    if checked_config != expected_config
        || manifest.units.is_empty()
        || manifest.units.len() != parsed.structured_source_units.len()
    {
        return Err(invalid("configuration or unit count mismatch"));
    }
    let images: BTreeMap<_, _> = parsed
        .images
        .iter()
        .map(|image| {
            (
                image.original_ref.clone(),
                hex::encode(Sha256::digest(&image.data)),
            )
        })
        .collect();
    if images.len() != parsed.images.len() || images != manifest.image_sha256 {
        return Err(invalid("image bytes differ from inventory receipt"));
    }
    let mut keys = BTreeSet::new();
    let mut pages = BTreeSet::new();
    let mut page_images = BTreeSet::new();
    for (ordinal, (entry, unit)) in manifest
        .units
        .iter()
        .zip(&parsed.structured_source_units)
        .enumerate()
    {
        let compatible_kind = match unit.kind {
            StructuredSourceUnitKind::Section => {
                matches!(entry.kind.as_str(), "paragraphs" | "section" | "pdf_page")
            }
            StructuredSourceUnitKind::TableRegion => entry.kind == "table",
            StructuredSourceUnitKind::ImageRegion => {
                entry.kind == "image" && entry.status == "not_checked"
            }
            StructuredSourceUnitKind::AttachmentRegion => {
                entry.kind == "not_checked" && entry.status == "not_checked"
            }
            _ => false,
        };
        if !compatible_kind || unit.ordinal as usize != ordinal {
            return Err(invalid("unit carrier kind or occurrence order mismatch"));
        }
        let outline = entry
            .heading_level
            .is_none_or(|level| entry.kind == "paragraphs" && (1..=9).contains(&level));
        let region = entry
            .field_region
            .as_deref()
            .is_none_or(|region| matches!(region, "toc" | "field"));
        if entry.unit_key != unit.key
            || !keys.insert(&entry.unit_key)
            || entry.part.is_empty()
            || entry.kind.is_empty()
            || !outline
            || !region
            || !matches!(entry.status.as_str(), "extracted" | "not_checked")
            || (entry.status == "not_checked")
                != entry
                    .reason
                    .as_ref()
                    .is_some_and(|reason| !reason.is_empty())
        {
            return Err(invalid("unit mapping or explicit coverage status invalid"));
        }
        if file_type == "pdf" {
            let page = match unit.locator {
                StructuredSourceLocator::Page { page_ordinal, .. }
                | StructuredSourceLocator::PageTable { page_ordinal, .. } => page_ordinal,
                StructuredSourceLocator::Image {
                    page_ordinal: Some(page),
                    ..
                } => page,
                _ => return Err(invalid("PDF unit has no physical page")),
            };
            if entry.part != format!("pdf:page:{}", page + 1) {
                return Err(invalid("PDF sidecar page mismatch"));
            }
            if matches!(unit.locator, StructuredSourceLocator::Page { .. }) {
                pages.insert(page);
            }
            if let StructuredSourceLocator::Image {
                ref original_ref,
                left: None,
                top: None,
                right: None,
                bottom: None,
                ..
            } = unit.locator
            {
                if !manifest.image_sha256.contains_key(original_ref) {
                    return Err(invalid("page image has no byte identity"));
                }
                page_images.insert(page);
            }
        }
    }
    if file_type == "pdf"
        && !manifest.page_count.is_some_and(|count| {
            count > 0
                && pages.len() == count as usize
                && pages.iter().copied().eq(0..count)
                && page_images == pages
        })
    {
        return Err(invalid("PDF physical page inventory incomplete"));
    }
    Ok(manifest)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{ImageRef, StructuredSourceUnit};

    fn blank_pdf() -> ReadResult {
        let mut parsed = ReadResult {
            structured_source_units: vec![StructuredSourceUnit {
                key: "page:0:section:0".into(), ordinal: 0,
                kind: StructuredSourceUnitKind::Section, text: String::new(),
                locator: StructuredSourceLocator::Page { page_ordinal: 0, left: None, top: None, right: None, bottom: None }, grid: None,
            }],
            metadata: HashMap::from([("output_inventory_manifest".into(), serde_json::json!({
                "schema_version":1,"profile":OUTPUT_INVENTORY_PROFILE,"file_sha256":"immutable",
                "parser":"profile/implementation/sha/pdfium/version","page_count":1,"image_sha256":{},
                "config":{"profile":OUTPUT_INVENTORY_PROFILE,"preserve_all_pages":true,"preserve_page_images":true,"text_cleaning":false,"settings":{}},
                "units":[{"unit_key":"page:0:section:0","part":"pdf:page:1","ordinal":0,
                    "kind":"pdf_page","bookmarks":[],"fields":[],"status":"extracted","reason":null}]
            }).to_string())]), ..Default::default()
        };
        parsed.images.push(ImageRef {
            original_ref: "images/page.png".into(),
            data: vec![1, 2, 3],
            ..Default::default()
        });
        parsed.structured_source_units.push(StructuredSourceUnit {
            key: "page:0:image:0".into(),
            ordinal: 1,
            kind: StructuredSourceUnitKind::ImageRegion,
            text: String::new(),
            grid: None,
            locator: StructuredSourceLocator::Image {
                original_ref: "images/page.png".into(),
                width: 10,
                height: 10,
                media_type: "image/png".into(),
                page_ordinal: Some(0),
                compound_parent: None,
                left: None,
                top: None,
                right: None,
                bottom: None,
            },
        });
        let mut receipt: Value =
            serde_json::from_str(&parsed.metadata["output_inventory_manifest"]).unwrap();
        receipt["image_sha256"] =
            serde_json::json!({"images/page.png":hex::encode(Sha256::digest([1,2,3]))});
        receipt["units"]
            .as_array_mut()
            .unwrap()
            .push(serde_json::json!({
                "unit_key":"page:0:image:0","part":"pdf:page:1","ordinal":1,"kind":"image",
                "bookmarks":[],"fields":[],"status":"not_checked","reason":"visual review required"
            }));
        parsed
            .metadata
            .insert("output_inventory_manifest".into(), receipt.to_string());
        parsed
    }

    #[test]
    fn explicit_profile_accepts_blank_page_without_fake_text() {
        let parsed = blank_pdf();
        assert!(parsed.markdown.is_empty());
        assert_eq!(
            validate_output_inventory(&parsed, "immutable", "pdf")
                .unwrap()
                .page_count,
            Some(1)
        );
    }

    #[test]
    fn old_service_wrong_source_and_incomplete_pages_fail_closed() {
        assert!(validate_output_inventory(&ReadResult::default(), "immutable", "pdf").is_err());
        let mut parsed = blank_pdf();
        assert!(validate_output_inventory(&parsed, "different", "pdf").is_err());
        parsed.structured_source_units[0].key = "orphan".into();
        assert!(validate_output_inventory(&parsed, "immutable", "pdf").is_err());
        let mut parsed = blank_pdf();
        parsed.structured_source_units[0].locator = StructuredSourceLocator::Page {
            page_ordinal: 1,
            left: None,
            top: None,
            right: None,
            bottom: None,
        };
        assert!(validate_output_inventory(&parsed, "immutable", "pdf").is_err());
    }

    #[test]
    fn unreceipted_or_changed_image_payload_cannot_enter_frozen_inventory() {
        let mut parsed = blank_pdf();
        validate_output_inventory(&parsed, "immutable", "pdf").unwrap();
        parsed.images[0].data[0] = 9;
        assert!(
            validate_output_inventory(&parsed, "immutable", "pdf")
                .unwrap_err()
                .0
                .contains("image bytes differ")
        );
        let mut parsed = blank_pdf();
        let mut receipt: Value =
            serde_json::from_str(&parsed.metadata["output_inventory_manifest"]).unwrap();
        receipt["units"].as_array_mut().unwrap().pop();
        parsed.structured_source_units.pop();
        parsed
            .metadata
            .insert("output_inventory_manifest".into(), receipt.to_string());
        assert!(validate_output_inventory(&parsed, "immutable", "pdf").is_err());
    }
}
